design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+as ARRAY or MAP from columns in KSQL streams or tables into multiple rows. This is the main
+motivation for proposing this feature.
+
+Table funtions would also be useful for other cases such as splitting a string into words, or any
+case where you need to go from a single row to zero or more, for example, unaggregating data or
+duplicating rows.
+
+Here's an example of exploding an array from a column into multiple rows:
+
+We have a stream containing a batch of sensor readings from a particular sensor. For further
+stream processing we want to split them out into separate messages:
+
+Given:
+
+```
+{sensor_id:12345 readings: [23, 56, 3, 76, 75]}
+```
+
+We want to get:
+
+```
+{sensor_id:12345 reading: 23}
+{sensor_id:12345 reading: 56}
+{sensor_id:12345 reading: 3}
+{sensor_id:12345 reading: 76}
+{sensor_id:12345 reading: 75}
+```
+
+This could be accomplished by the following CSAS:
+
+```
+CREATE STREAM exploded_stream AS
+    SELECT sensor_id, EXPLODE(readings) AS reading FROM batched_readings;
+```
+
+Note that any scalar values in the select will be duplicated for each row in the output.
+
+
+## What is in scope
+
+We would like to implement
+
+* Table functions on the SELECT clause only
+* Multiple table functions on the SELECT clause
+* A set of built-in table functions such as EXPLODE (etc)
+* An ability for users to provide their own User Defined Table Functions (UDTFs) using annotations
+in a similar way to how UDFs and UDAFs are currently supported.
+
+## What is not in scope
+
+* Table functions on the FROM clause. We could allow table functions to take the place of other streams
+or tables in the FROM clause. We do not think there is much value in doing this as the main driver
+for this KLIP is in supporting `explode` functionality which is more simply implemented on the
+SELECT clause.
+* Table functions that output more than one column. Some databases support the flattening of an array
+ of N STRUCT types into N rows of M columns, where there is one column for each field of the struct.
+ Instead we propose that flattening an array of N structs will result in N rows with a single column
+ of type STRUCT. The fields of the struct can then be extracted if necessary using a further CSAS
+ which pulls out the struct fields.
+
+## Value/Return
+
+This functionality will make it easier for users to make end-end streaming apps using KSQL
+alone without having to have custom stages using Kafka Streams.
+
+We believe its a common case for users to have array or map like data structures as values
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
```

Comment:
it'd be nice to spell this out a little bit as there might be interesting consequences based on how we choose to model it. e.g. if it returns an `Iterable<TYPE>` vs a `List<TYPE>` there may be memory/performance implications we need to consider (e.g. the Iterable approach would open the door to lazy computation)

There's also a consideration about infinite returns. Do we want to support that? e.g. if we have an iterable instead of a list or an array, then a user could implement something like `RANDOMS()` which will infinitely output rows for one input.

Reply:
Hey @agavra Thank you for raising such an interesting point!

Sorry in advance for the long post, but this is one of my specialist areas.

In an ideal world I would love us to be able to provide functions that produce infinite results, or produce results asynchronously or lazily-  this opens us up to a lot more interesting use cases that could be implemented.

However for this to work it needs support from the underlying streams library. AIUI, right now the Kafka Streams DSL (fFlatMap) uses a synchronous blocking approach (i.e. “please give me the entire result of the FlatMap and I will block waiting for it).

See https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValues.java#L40

I don't think that's going to work for slow, infinite or async functions. It also doesn’t seem obvious how we could write our own custom Kafka Streams processor to do this.

To make a really scalable event processing platform based on the pull model (polling) you need some way of ensuring that calls to getNext() (e.g. on an iterator) never block. Blocking kills performance as the blocking thread can’t do other work while it’s blocked. For a timed iterator that only creates, say, one new result every 10 seconds, this would be a disaster. When there are many running queries in the server blocking, you’d end up with large number of threads just to keep the system responsive, plus RAM overhead per thread, not to mention overhead from extra context switching.

To avoid blocking on a poll(), scalable event processing systems (Rust futures are a good example of this) need some way of asking the task whether it’s ready or not so a poll doesn’t block, e.g. an isReady() method. If isReady() returns false, then the task needs to be parked, and the task also needs to wake itself up sometime in the future and reschedule itself when it has more data to provide. To cut a long story short, it gets complicated! A simple iterator is not going to help us here, the API of the function would need to support the non blocking semantics and it also needs these concepts built into the underlying stream processing framework. AIUI the blocking architecture of Kafka Streams is really not compatible with that way of working.

An alternative way of implementing this scalable is using an async model (push model). In this model there is no polling happening to see if results are ready. Instead results are provided asynchronously to the caller (usually via a callback function or a future) when they are ready. In this model there is no blocking either so it can scale well, but it also requires support from the underlying streams framework to work. With the push model, it also needs to handle back pressure (aka flow control), e.g. to prevent very fast producers of messages overwhelming things downstream and causing OOMs. This model is used quite widely including Vert.x, ReactiveX, Spring Reactor etc.

To summarise: I would love slow/infinite/async functions, but I don’t see any way of implementing these efficiently/scalably using Kafka Streams using the DSL or Processor APIs (perhaps there is another API I am not aware of?) And even if we’re not tied to Kafka Streams the choice of API for the function would depend on whether the underlying streams framework uses a push or pull model.

Reply:
I may be wrong, as I'm not KS expert, but I think having a infinite UDTF would be possible with the Processor API of KS, as it does not expect all the results in one go, (unlike `flatmap`), instead you can call `ProcessorContext#forward(k, v)` as often as you like.

However, I'm not sure how gracefully KS would handle such a case. I have a feeling any infinite UDTF would cause the KS consumers to be ejected from the consumer group as they can't poll Kafka until the UDTF returns.

cc @guozhangwang @mjsax  @bbejeck  to confirm.

That said, does it actually make much sense to use an infinite UDTF on a table/stream, i.e. topic, source? Essentially, only the first record per-partition would be read, and then the infinite output would commence. As consumers were ejected from their groups those partitions may be assigned to stream threads which previously had no tasks, (threads with tasks will be stuck output infinite rows, so won't have new tasks assigned). As the partition moved to another stream thread it too would read the first message and commence infinite output.  So, the result would be inconsistent as we could have more than one stream thread performing infinite output per source partition.

That is not to say that infinite UDTFs are not of use, only that they don't make sense on a infinite source.  Maybe they do make sense on a finite source, e.g. using Oracle's `DUAL` source equivalent:

```sql
SELECT myInfiniteUdtf("blah') FROM DUAL;
```

Or using another UDTF to generate a finite source:

```sql
-- kick off datagen for the three types provided.
SELECT myDataGen(type) FROM explode("users", "clicks", "status-codes") AS type;
```

I'm not suggesting we support all of this / any of this now, though we can think about if we want to. Regardless, we should at least consider if our API can be designed, as @agavra says, to support such a thing in the future. 

As for the comments about async/sync push/pull threading etc.  These are obviously all valid points, but separate to the design of our API for UDTFs.  Having a thread per infinite/slow/async UDTF may be inefficient, but we can always later extend the framework from, say, UDTFs that return `Iterator<String>` to `Iterator<Future<String>>` to allow us to handle such requests more efficiently.

Reply:
@big-andy-coates I brought up infinite UDTFs more as a way of demonstrating how the API can be used (or maybe abused) if we designed it in one way vs another, so I think the discussion is important when we are designing our APIs. If we decide that infinite (or more specifically, anything that's non-immediate) UDTFs are valuable, then we need to support some iterable-type (as @purplefox pointed out, `Iterator` might not be smart) interface as opposed to one that's set in stone.

With regards to whether or not it's possible, that's also something I'm not familiar enough to comment on (waiting for the streams folk to chime in here), but I think if we decided that it's something that we want we should design our API around that, potentially updating KS to support it later.

Reply:
> I may be wrong, as I'm not KS expert, but I think having a infinite UDTF would be possible with the Processor API of KS, as it does not expect all the results in one go, (unlike `flatmap`), instead you can call `ProcessorContext#forward(k, v)` as often as you like.

I am also not a KS expert but my understanding is there would be a thread blocking in KS until all the results from the infinite or slow UDTF are returned, and for infinite stream that never occurs. That doesn't seem very scalable.

> That is not to say that infinite UDTFs are not of use, only that they don't make sense on a infinite source. 

Right I don't think the suggestion was for them to work on an infinite source, more like things that can create a stream of random numbers of timing events (a tick every 10 seconds for example).

> As for the comments about async/sync push/pull threading etc. These are obviously all valid points, but separate to the design of our API for UDTFs. Having a thread per infinite/slow/async UDTF may be inefficient, but we can always later extend the framework from, say, UDTFs that return `Iterator<String>` to `Iterator<Future<String>>` to allow us to handle such requests more efficiently.

That API very much depends on what approach (push/pull) is taken. With a polling approach it would also be necessary to signal readiness, park and reschedule tasks - it's complex! For an async approach a simple future probably wouldn't suffice as we would need some way of propagating back pressure. All these considerations would need to be built into the API.

Reply:
The current threading model of KS would indeed block the consumer from calling "poll" before a record has "completely" traversed through the topology, i.e. all its inferred records from the operator has been processed.

However, Streams team have some discussions about refactoring the threading model https://confluentinc.atlassian.net/wiki/spaces/KSTREAMS/pages/893749214/2019-06-07+and+more+Deep+Dive+on+Dynamic+Scaling, the main goal is to saturate the resources of a single node, but as a by-product it also means async-processing from the pov of the main thread calling poll, so infinite results it would not be a problem for blocking, BUT they would still be tricky from the committing aspect, as you still have to decide when it is safe to commit that "all records till this boundary has been processed".

Reply:
Thanks for the insight @guozhangwang I will follow the discussion on the KS threading model changes :)